src/target/riscv/riscv-013.c

```diff
@@ -4394,182 +4394,299 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
+	if (execute_abstract_command(target, command) != ERROR_OK)
 		return ERROR_FAIL;
 
-	/* s0 holds the next address to write to
-	 * s1 holds the next data value to write
+	/* Address on the target was incremented during the execution.
 	 */
+	*address_p += size;
 
-	int result = ERROR_OK;
-	if (riscv_save_register(target, GDB_REGNO_S0) != ERROR_OK)
+	return dm_write(target, DM_ABSTRACTAUTO,
+			1 << DM_ABSTRACTAUTO_AUTOEXECDATA_OFFSET);
+}
+
+/**
+ * This function attempts to restore the pipeline after a busy on abstract
+ * access or a DMI busy by reading the content of s0 -- the address of the
+ * failed write.
+ */
+static int write_memory_progbuf_inner_on_busy(struct target *target,
+		target_addr_t *address, uint32_t size, const uint8_t *buffer)
+{
+	riscv013_clear_abstract_error(target);
+	increase_ac_busy_delay(target);
+
+	if (dm_write(target, DM_ABSTRACTAUTO, 0) != ERROR_OK)
 		return ERROR_FAIL;
-	if (riscv_save_register(target, GDB_REGNO_S1) != ERROR_OK)
+
+	target_addr_t address_on_target;
+	if (register_read_direct(target, &address_on_target, GDB_REGNO_S0) != ERROR_OK)
 		return ERROR_FAIL;
+	const uint8_t * const curr_buff = buffer + (address_on_target - *address);
+	LOG_TARGET_DEBUG(target, "Restarting from 0x%" TARGET_PRIxADDR, *address);
+	*address = address_on_target;
+	return write_memory_progbuf_inner_startup(target, address, curr_buff, size);
+}
 
-	/* Write the program (store, increment) */
-	struct riscv_program program;
-	riscv_program_init(&program, target);
-	if (riscv_enable_virtual && has_sufficient_progbuf(target, 5) && get_field(mstatus, MSTATUS_MPRV))
-		riscv_program_csrrsi(&program, GDB_REGNO_ZERO, CSR_DCSR_MPRVEN, GDB_REGNO_DCSR);
+/**
+ * This function fills the batch with writes and returns the address on target
+ * of the next ellement, if the batch is executed fully.
+ */
+static target_addr_t write_memory_progbuf_inner_fill_batch(struct riscv_batch *batch,
+		target_addr_t start_address, target_addr_t end_address, uint32_t size,
+		const uint8_t *buffer)
+{
+	assert(size <= 8);
+	const uint32_t reads_per_element = size > 4 ? 2 : 1;
+	const uint32_t batch_capacity = riscv_batch_available_scans(batch) / reads_per_element;
+	const target_addr_t distance = MIN((target_addr_t)batch_capacity * size,
+			end_address - start_address);
 
-	switch (size) {
-		case 1:
-			riscv_program_sbr(&program, GDB_REGNO_S1, GDB_REGNO_S0, 0);
-			break;
-		case 2:
-			riscv_program_shr(&program, GDB_REGNO_S1, GDB_REGNO_S0, 0);
-			break;
-		case 4:
-			riscv_program_swr(&program, GDB_REGNO_S1, GDB_REGNO_S0, 0);
-			break;
-		case 8:
-			riscv_program_sdr(&program, GDB_REGNO_S1, GDB_REGNO_S0, 0);
-			break;
-		default:
-			LOG_TARGET_ERROR(target, "Unsupported size: %d", size);
-			result = ERROR_FAIL;
-			goto error;
+	for (target_addr_t address = start_address;
+			address - start_address < distance;
+			address += size, buffer += size) {
+		uint64_t value = get_from_buffer_and_log_write(address, buffer, size);
+		if (size > 4)
```

Comment:
Really minor detail. If you like:

```suggestion
		if (writes_per_element == 2)
```

Reply:
Addressed.